state/store.go

```diff
@@ -212,133 +233,172 @@ func (store dbStore) Bootstrap(state State) error {
-			err = batch.Delete(calcConsensusParamsKey(h))
-			if err != nil {
-				return err
-			}
+}
+
+// pruneConsensusParams calls an iterator from base height to retain height batch deleting 
+// all consensus params in between. If the consensus params at the new base height is dependent
+// on a prior height then this will keep that lower height to. 
+func (store dbStore) pruneConsensusParams(height int64) error {
+	paramsInfo, err := store.loadConsensusParamsInfo(height)
+	if err != nil {
+		return fmt.Errorf("consensus params at height %v not found: %w", height, err)
+	}
+
+	// As we don't save the consensus params at every height, only when there is a consensus params change,
+	// we must not prune (or save) the last consensus params that the consensus params info at height
+	// is dependent on.  
+	var lastRecordedConsensusParams *tmstate.ConsensusParamsInfo
+	if paramsInfo.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
+		lastRecordedConsensusParams, err = store.loadConsensusParamsInfo(paramsInfo.LastHeightChanged)
+		if err != nil ||  lastRecordedConsensusParams.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
+			return fmt.Errorf(
+				"couldn't find consensus params at height %d as last changed from height %d: %w",
+				paramsInfo.LastHeightChanged,
+				height,
+				err,
+			)
 		}
+	}
 
-		err = batch.Delete(calcABCIResponsesKey(h))
+	// batch delete all the consensus params up to height
+	if err := store.batchDelete(consensusParamsKey, height); err != nil {
+		return err
+	}
+
+	// check if we had a dependent consensus params info of an earlier height that we need to save.
+	// If so then we restore it now
+	if !lastRecordedConsensusParams.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
+		bz, err := lastRecordedConsensusParams.Marshal()
 		if err != nil {
 			return err
 		}
-		pruned++
 
+		if err := store.db.Set(consensusParamsKey(paramsInfo.LastHeightChanged), bz); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// pruneABCIResponses calls an iterator from base height to retain height batch deleting 
+// all abci responses in between
+func (store dbStore) pruneABCIResponses(height int64) error {
+	return store.batchDelete(abciResponsesKey, height)
+}
+
+// batchDelete is a generic function for deleting a range of values based on the lowest
+// height up to but excluding retainHeight
+func (store dbStore) batchDelete(key func(int64) []byte, retainHeight int64) error {
+	iter, err := store.db.Iterator(
+		key(1),
+		key(retainHeight),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer iter.Close()
+
+	batch := store.db.NewBatch()
+	defer batch.Close()
+
+	pruned := 0
+	for ; iter.Valid(); iter.Next() {
+		if err := batch.Delete(iter.Key()); err != nil {
+			return fmt.Errorf("pruning error at height %d: %w", decodeKey(iter.Key()), err)
+		}
+
+		pruned++
 		// avoid batches growing too large by flushing to database regularly
-		if pruned%1000 == 0 && pruned > 0 {
+		if pruned%1000 == 0 {
 			err := batch.Write()
```

Comment:
@erikgrinaker Eveything but the last case of `TestPruneStates` is passing. However, when I prune more than 1000 states it seems to hold indefinitely. I mucked around a bit and found that this is the exact line that it holds on. When I compare with master, this has the same coding pattern so I'm a bit lost as to why it's doing it now. Do you have an idea what it could be?

Reply:
You're modifying the key domain while you're iterating over it. tm-db's `Iterator` documentation says:

> No writes can happen to a domain while there exists an iterator over it, some backends may take out database locks to ensure this will not happen.

This is an annoying limitation, but it is database-dependent and not all embedded databases can handle this.

My guess is that you're getting a deadlock because the iterator is already holding a lock on the database, and the batch write is trying to obtain that lock (which won't get released until the iterator is closed).

Two options:

1. Pregenerate the entire list of keys first, close the iterator, and then delete the keys (in batches of 1000).

2. Close the iterator before writing the batch, making a note of the next upcoming key, and then resume iteration after the batch is written.

I'd probably go for 2, since it uses a bounded amount of memory (1 can consume all memory on the system depending on the size of the database).

Reply:
Right that makes sense. Yeah I will implement the 2nd one

Reply:
Actually, you wouldn't even need to keep track of the next key - we're always pruning from 0, so we can just restart iteration from 0 every time (which will start at the next existing key).

Reply:
Yess I got it to work now :pray:  thanks 

I found it simpler to just recurse seeing, as you said, the iterator range is the same. This won't be a problem for the call stack right? Hmm actually maybe if we calling this over millions of blocks

Reply:
Hm, recursion is an interesting approach here - couldn't we just use an outer while loop, i.e. `for iter.Valid() { }`?

Reply:
yeah I decided to remove it